packages/abi-utils/lib/arbitrary.ts

```diff
@@ -0,0 +1,315 @@
+            ? { stateMutability }
+            : {
+                stateMutability,
+                payable: stateMutability === "payable",
+                constant:
+                  stateMutability === "view" || stateMutability === "pure"
+              }
+        )
+    )
+    .map(records => records.reduce((a, b) => ({ ...a, ...b }), {}));
+
+export const ConstructorEntry = () =>
+  fc
+    .tuple(
+      fc.record({
+        type: fc.constant("constructor"),
+        inputs: fc.array(Parameter())
+      }),
+      fc
+        .tuple(
+          fc.oneof(fc.constant("nonpayable"), fc.constant("payable")),
+          fc.boolean()
+        )
+        .map(([stateMutability, includeOptionals]) =>
+          !includeOptionals
+            ? { stateMutability }
+            : {
+                stateMutability,
+                payable: stateMutability === "payable",
+                constant:
+                  stateMutability === "view" || stateMutability === "pure"
+              }
+        )
+    )
+    .map(records => records.reduce((a, b) => ({ ...a, ...b }), {}));
+
+export const Abi = () =>
+  fc
+    .tuple(
+      ConstructorEntry(),
+      FallbackEntry(),
+      ReceiveEntry(),
+      fc.array(fc.oneof(FunctionEntry(), EventEntry()))
+    )
+    .chain(([constructor, fallback, receive, entries]) =>
+      fc.shuffledSubarray([constructor, fallback, receive, ...entries])
+    );
+
+namespace Numerics {
+  // 0 < n <= 32
+  export const Bytes = () => fc.nat(31).map(k => 32 - k);
+
+  // o < n <= 256
+  export const Bits = () => Bytes().map(k => 8 * k);
+
+  // 0 < n < 80
+  export const DecimalPlaces = () => fc.nat(79).map(k => k + 1);
+
+  export const Precision = () => fc.tuple(Bits(), DecimalPlaces());
+}
+
+namespace Primitives {
+  export const Uint = () => Numerics.Bits().map(m => `uint${m}`);
+  export const Int = () => Numerics.Bits().map(m => `int${m}`);
+  export const Address = () => fc.constant("address");
+  export const Bool = () => fc.constant("bool");
+  export const Fixed = () =>
+    Numerics.Precision().map(([m, n]) => `fixed${m}x${n}`);
+  export const Ufixed = () =>
+    Numerics.Precision().map(([m, n]) => `ufixed${m}x${n}`);
+  export const BytesM = () => Numerics.Bytes().map(m => `bytes${m}`);
+  export const Function = () => fc.constant("function");
+  export const Bytes = () => fc.constant("bytes");
+  export const String = () => fc.constant("string");
+  export const Tuple = () => fc.constant("tuple");
+}
+
+const Primitive = () =>
+  fc.oneof(
+    Primitives.Uint(),
+    Primitives.Int(),
+    Primitives.Address(),
+    Primitives.Bool(),
+    Primitives.Fixed(),
+    Primitives.Ufixed(),
+    Primitives.BytesM(),
+    // Primitives.Function(), // TODO re-add
+    Primitives.Bytes(),
+    Primitives.String(),
+    Primitives.Tuple()
+  );
+
+const Type: fc.Memo<string> = fc.memo(n =>
+  n === 0
+    ? Primitive()
+    : fc.oneof(Primitive(), ArrayFixed(n > 3 ? 3 : n), ArrayDynamic(n))
```

Comment:
The fact that `ArrayFixed` has its dimension capped warrants at least a comment explaining why.